Automattic/jetpack
# Updates to disconnect modal to improve accessibility

#### Changes proposed in this Pull Request:

* This PR Makes several updates to the disconnect flow to better align with WCAG accessibility guidelines. These changes are outlined below, organized by each step of the flow. Screenshots of these steps are included in the test plan.

**Changes that apply to the flow in general**
* The `<Modal />` component has been given a more consistent accessible label using the `contentLabel` prop. Previously it used `aria-labelledby` and referenced the step headline, which changed from step to step.
* The modal can now be closed by pressing the escape key or by clicking outside the modal. This is not necessarily a requirement, but it is more common/ expected behavior for modals. Depending on whether or not the site is disconnected, dismissing the modal has different effects. Additionally, the modal cannot be dismissed at all if the site is actively disconnecting.
* The action buttons now have a slight hover state 
* Link-style buttons now have a noticeable focus state
* Link-style buttons that are functioning as buttons now use `<button>` elements instead of `<a>` elements

**Disconnect Step**
* Cards that show other connected plugins and benefits of the Jetpack plugin have been refactored as unordered lists to be more semantically valuable.
* Blue links in the list of generic Jetpack benefits have an underline to provide sufficient differentiation of links from surrounding body text.
* The Disconnect button uses the `aria-live` attribute to announce "Disconnecting..." when using a screen reader.

**Disconnect Confirmation Step**
* This step uses the `aria-live` attribute to read the content of this screen when it appears after disconnection (announcing dynamic content).

**Survey Step**
* This step uses the `aria-live` attribute to read the content of this screen when it appears (announcing dynamic content).
* The survey questions have been refactored to use a series of radio inputs within a `<form>` element to be more semantically correct for the function of the control (only one option can be selected at a time)
* Hover, focus and selected states of the survey options have been updated to (hopefully) make it easier to understand when a choice is selected.

**Thank You Step**
* This step uses the `aria-live` attribute to read the content of this screen when it appears after disconnection (announcing dynamic content).

#### Does this pull request change what data or activity we track or use?


No.

#### Testing instructions:






Most of these changes have an impact when testing with a screen reader and/ or using the keyboard to navigate the disconnection flow. For MacOS, the built in VoiceOver screen reader is a good one to test with, and can be activated by pressing `command + F5`. If you have not used a screen reader before, you may need to take a bit of time to get familiar with how to navigate with it. These test instructions will make note of how the screen reader should announce certain elements on the page and how the keyboard can be used to navigate.

* Check out/ apply this branch to your Jetpack test site
* Make sure Jetpack is connected and navigate to the Jetpack dashboard.
* From the Jetpack dashboard, click on "Manage Connection" to open the disconnect flow.
![Screen Shot 2021-12-09 at 5 08 47 PM](https://user-images.githubusercontent.com/18016357/145487922-156f6a37-c7e0-4cfb-abad-cdd4753c8d92.png)

* If using a screen reader, it should announce "Disconnect Jetpack Dialog" when the modal opens
* Note, you should be able to press the escape key or click outside the modal to close it.
* Navigating through the content with a screen reader, the list of plugins and benefits provided by Jetpack should be identified as lists.
* Using the tab key, you should be able to focus on all links and buttons. All links and buttons should have noticeable focus states.
* Press "Enter" or "Spacebar" when focused on the Disconnect button to initiate a site disconnect. If using a screen reader, it should announce "Disconnecting" when the disconnection request is made.
* After the site disconnects, you will be presented with the disconnect confirmation screen.
![Screen Shot 2021-12-09 at 5 09 29 PM](https://user-images.githubusercontent.com/18016357/145488559-042cc180-e422-4c88-88e0-a78a189ffbea.png)

* If using a screen reader, the content of this screen should be read aloud when it appears.
* You can use the tab key to navigate to the different options on the screen. All links and buttons should have noticeable focus states.
* Press "Enter" or "Spacebar" when focused on the "Help us improve" button to navigate to the survey step.
![Screen Shot 2021-12-09 at 5 10 47 PM](https://user-images.githubusercontent.com/18016357/145488795-1afa0276-e755-4e30-82b7-e73e8fae81e3.png)

* If using a screen reader, the content of the survey screen should be read aloud when it appears.
* You can use the tab key to navigate between the survey option set, the custom input field, and the buttons below the survey.
* Because the survey options are managed as a set of radio buttons, you can use the arrow keys to move between and select one of the options. If focus leaves and returns to the set of survey responses, focus will be given to the response that has already been selected. Try navigating and selecting one of the responses using the arrow keys.
* The last survey option has a field for custom input. If input is entered into this field, this last "Other" option is automatically selected as the active option.
* Confirm that pressing "Enter" while focused on the survey options does not submit the survey.
* After selecting a survey option, use the tab key on the keyboard to navigate to the "Submit Feedback" button and press "Enter" to submit the survey.
* After the survey is submitted, you will be presented with the "Thank You" screen:
![Screen Shot 2021-12-09 at 5 11 33 PM](https://user-images.githubusercontent.com/18016357/145489552-d6e9e929-7d30-495f-a734-69a9afb188cc.png)

* If you are using a screen reader, the content of the thank you screen should be read aloud when it appears.
* You can then use the tab key on your keyboard to focus on the "Back to my Website" button. Press enter on this button to close the modal and exit the flow.
